### PR TITLE
sql: Fix incorrect handling of types

### PR DESCRIPTION
**Pull request by shmsr**
## Proposed commit message

This fixes the incorrect handling of types in SQL module. More details on the issue are present here: https://github.com/elastic/beats/issues/40090

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [x] Test with Elasticsearch
- [x] Test if valid inputs are treated correctly
- [x] Test if there are other cases too

## How to test this PR locally

As usual.

## Related issues

- https://github.com/elastic/beats/issues/40090

### CODE REVIEW

File: x-pack/metricbeat/module/sql/query/query.go
Comment (shmsr) [inline x-pack/metricbeat/module/sql/query/query.go:348]: This is where based on types, key-vals are assigned under `numeric`, `string` or `bool`

File: metricbeat/helper/sql/sql.go
Comment (ishleenk17) [inline metricbeat/helper/sql/sql.go:175]: For something like below: 
data := []byte("123.45")

Should data be a string or numeric. 
In previous code we were changing to float, now it will always be a string. 
What is the expectation of the SQL module for this ?

Comment (shmsr) [reply metricbeat/helper/sql/sql.go:175]:
```
integer types smallint, integer, and bigint are returned as int64
floating-point types real and double precision are returned as float64
character types char, varchar, and text are returned as string
temporal types date, time, timetz, timestamp, and timestamptz are returned as time.Time
the boolean type is returned as bool
the bytes type is returned as []byte
```

This is how SQL types are converted to Go types using the library for Postgres.

If you see if someone has set a type as BYTES in SQL, there's expectation to keep it as []byte or string. I am not sure why earlier there was a decision to convert it to float as this as issues:

- If the query returns []byte equivalent of "00100"; converting it to float is possible and it would truncate the leading 0s. And no way to get it back.

I think it's better to keep it as string and not change it to float from the code. If someone needs to convert it to float then they can use ingest pipelines; which is doable but not the other way around i.e., float to string.

In our case, for which the bug was filed. field was a VARCHAR and Postgres client library converts it to Go's string type; and it should be kept string only (no auto assumption and conversion should be done, imo). If required, they can always do it in their custom ingest pipelines.

Comment (ishleenk17) [reply metricbeat/helper/sql/sql.go:175]: This could change the behaviour for the users using the module for [ ] byte datatype. As they might be expecting numeric and might have added their own customizations per that. 

But, IMO also, its ok to have it as string. 
Just considering it should not break for the existing users.

**Review by ishleenk17 — APPROVED** [review]

File: metricbeat/helper/sql/sql.go
Comment (shmsr) [reply metricbeat/helper/sql/sql.go:190]: Yes, I'll explain this why I've kept it like this:

Check this: https://pkg.go.dev/github.com/lib/pq#hdr-Data_Types

```
integer types smallint, integer, and bigint are returned as int64
floating-point types real and double precision are returned as float64
character types char, varchar, and text are returned as string
temporal types date, time, timetz, timestamp, and timestamptz are returned as time.Time
the boolean type is returned as bool
the bytes type is returned as []byte
```

Although in the library they just talk about a few types, I didn't found doc for the others types and what they are converted to in Go.

For example, see: https://www.postgresql.org/docs/current/datatype-numeric.html

See there are types like `decimal`, and there are even lot more types: https://www.postgresql.org/docs/current/datatype.html (e.g., inet, cidr, etc.).

So logic behind the `default` case it to handle them for which I do not know types. So kept the logic as before i.e.,

Convert the value to equivalent string and iff it is convertable to float, do it else keep it a string. But added logic here is to handle the cases where leading zeroes could be there; i.e., it is convertable to float but truncation of leading 0s will happen. So, logic is finally:

> convert to string -> if it has leading zeroes (not immediately followed by a dot), keep it a string -> check if convertable to float; if yes, do it or else keep it a string

It's just as before but sightly better. But you are correct, we addressed the actual bug in the long case where we have already the `string`.

I do not assume that users will have to deal with default case unless they are using those special types for which we do not have proper doc in our client library for Postgres.

**Comment by shmsr** [comment]
@AndersonQ / @faec Could you please review this PR? This is an important bugfix that requires immediate attention. Thank you for your help.

We are planning to get this merged ASAP to make it available for next release.

**Review by devamanv — APPROVED** [review]

**Review by cmacknz — APPROVED** [review]
I have only skimmed this, approving to unblock.

Probably you could adjust the CODEOWNER so the data plane team isn't ping for this change next time.

File: metricbeat/helper/sql/sql.go
Comment (shmsr) [reply metricbeat/helper/sql/sql.go:169]: Yes, I agree with your point. There are many code smells that need to be addressed. I'll create a task and add it to our backlog.

**Pull request merged**
